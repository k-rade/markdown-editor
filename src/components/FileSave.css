.file-save {
    position: relative;
}

.file-save-dialog hr {
    border-top: 1px solid var(--color-gray-900);
}

.file-save-dialog {
    left: 3rem;
    top: -0.5rem;
    border-radius: 6px;
    position: absolute;
    border: 1px solid var(--color-gray-500);
    background-color: var(--color-gray-700);
    box-shadow: 0 10px 16px var(--color-black-three), 0 0 1px var(--color-black-two) inset;
}

.file-save-dialog::before {
    z-index: 1;
    top: 1rem;
    left: -9px;
    position: absolute;
    content: "";
    width: 10px;
    height: 20px;
    background-color: var(--color-gray-500);
    clip-path: polygon(0 50%, 100% 100%, 100% 0);
}

.file-save-dialog::after {
    z-index: 2111;
    top: 17px;
    left: -7px;
    position: absolute;
    content: "";
    width: calc(10px - 2px);
    height: calc(20px - 2px);
    background-color: var(--color-gray-700);
    clip-path: polygon(0 50%, 100% 100%, 100% 0);
}

.file-save-dialog-header {
    padding: 0.5rem 1rem;
    display: flex;
    align-items: center;
    justify-content: space-between;
    font-weight: 500;
}

.file-save-dialog-header .icon {
    width: 1.25rem;
    height: 1.25rem;
}


.file-name-input {
    position: relative;
    margin: 0 0 0.5rem;
}

.file-name-input-field {
    color: var(--color-white);
    padding: 0.5rem 1rem;
    border-radius: 6px;
    outline: none;
    border: none;
    background-color: var(--color-black-three);
    margin: 1rem 1rem 1.25rem;
    font-family: var(--font);
    caret-color: var(--color-primary-500);
}

.file-name-error {
    bottom: 0;
    left: 1.25rem;
    white-space: nowrap;
    position: absolute;
    font-size: 0.75rem;
    color: var(--color-primary-500);
}

.file-save-dialog-buttons {
    display: flex;
    gap: 0.5rem;
    padding: 1rem;
}