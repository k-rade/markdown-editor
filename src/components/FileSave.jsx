import { useState } from "react";
import Button from "./Button";
import IconButton from "./IconButton";
import { ReactComponent as SaveIcon } from "../assets/icons/save.svg";
import { ReactComponent as CloseIcon } from "../assets/icons/x.svg";

import "./FileSave.css";

const FileSave = ({ saveFile, fileName, changeFileName }) => {
  const [state, setState] = useState({
    showDialog: false,
    error: "",
  });

  const handleValidation = (event) => {
    changeFileName(event.target.value);
    if (!event.target.value) {
      setState({
        ...state,
        error: "Must not be blank",
      });
    } else {
      setState({ ...state, error: "" });
    }
  };

  const handleFileSave = () => {
    saveFile();
    setState({ showDialog: false, error: "" });
  };

  return (
    <div className="file-save">
      <IconButton
        icon={<SaveIcon />}
        onClick={() =>
          setState({ ...state, showDialog: !state.showDialog, error: "" })
        }
      />
      {state.showDialog && (
        <div className="file-save-dialog">
          <div className="file-save-dialog-header">
            <div>Name your file</div>
            <IconButton
              icon={<CloseIcon />}
              onClick={() => setState({ showDialog: false, erorr: "" })}
            />
          </div>
          <div className="file-name-input">
            <input
              placeholder="File name"
              className="file-name-input-field"
              type="text"
              value={fileName}
              onChange={handleValidation}
            />
            {state.error && (
              <span className="file-name-error">{state.error}</span>
            )}
          </div>
          <div className="file-save-dialog-buttons">
            <Button
              disabled={!fileName}
              label="Save"
              variant="primary"
              onClick={handleFileSave}
            />
            <Button
              label="Cancel"
              variant="ghost"
              onClick={() => setState({ showDialog: false, error: "" })}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default FileSave;
