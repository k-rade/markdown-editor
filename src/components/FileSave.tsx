import {
  ChangeEvent,
  EffectCallback,
  useEffect,
  useRef,
  useState,
} from "react";
import Button from "./Button";
import IconButton from "./IconButton";
import { ReactComponent as CloseIcon } from "../assets/icons/x.svg";
import { ReactComponent as PlusIcon } from "../assets/icons/save.svg";

import "./FileSave.css";

interface Props {
  fileName: string;
  saveFile: () => void;
  changeFileName: (fileName: string) => void;
}

const FileSave = ({
  saveFile,
  fileName,
  changeFileName,
}: Props): JSX.Element => {
  const dialog = useRef<any>();
  const [state, setState] = useState<{ showDialog: boolean; error: string }>({
    showDialog: false,
    error: "",
  });

  const handleValidation = (event: ChangeEvent<HTMLInputElement>): void => {
    changeFileName(event.target.value);
    if (!event.target.value) {
      setState({
        ...state,
        error: "Must not be blank",
      });
    } else {
      setState({ ...state, error: "" });
    }
  };

  const handleFileSave = (): void => {
    saveFile();
    setState({ showDialog: false, error: "" });
  };

  useEffect((): ReturnType<EffectCallback> => {
    const handleClickOutside = (event: Event): void => {
      if (dialog.current && !dialog.current.contains(event.target)) {
        setState({ showDialog: false, error: "" });
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return (): void =>
      document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  return (
    <div className="file-save" ref={dialog}>
      <IconButton
        icon={<PlusIcon />}
        onClick={() =>
          setState({ ...state, showDialog: !state.showDialog, error: "" })
        }
        tooltip="Save file"
      />
      {state.showDialog && (
        <div className="file-save-dialog">
          <div className="file-save-dialog-header">
            Name File
            <IconButton
              icon={<CloseIcon />}
              onClick={() => setState({ showDialog: false, error: "" })}
            />
          </div>
          <div className="file-save-dialog-input-wrapper">
            <input
              placeholder="File name"
              className="file-save-dialog-input-field"
              type="text"
              value={fileName}
              onChange={handleValidation}
            />
            {state.error && (
              <span className="file-save-dialog-input-error">
                {state.error}
              </span>
            )}
          </div>
          <div className="file-save-dialog-button">
            <Button
              disabled={!fileName}
              label="Save"
              onClick={handleFileSave}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default FileSave;
