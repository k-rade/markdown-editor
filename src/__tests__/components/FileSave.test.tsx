import { fireEvent, render } from "@testing-library/react";
import FileSave from "../../components/FileSave";

describe("FileSave", (): void => {
  it("should render", () => {
    const { baseElement } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={jest.fn()}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("should render with popup", (): void => {
    const { getByText, baseElement } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={jest.fn()}
      />
    );

    fireEvent.click(getByText("save.svg"));

    expect(baseElement).toMatchSnapshot();
  });

  it("should close file save popup", (): void => {
    const { getByText, queryByTestId } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={jest.fn()}
      />
    );

    fireEvent.click(getByText("save.svg"));

    expect(queryByTestId("file-save-dialog")).toBeInTheDocument();

    fireEvent.click(getByText("x.svg"));

    expect(queryByTestId("file-save-dialog")).not.toBeInTheDocument();
  });

  it("should close file save popup when clicked outside", (): void => {
    const { getByText, queryByTestId } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={jest.fn()}
      />
    );

    fireEvent.click(getByText("save.svg"));

    expect(queryByTestId("file-save-dialog")).toBeInTheDocument();

    fireEvent.mouseDown(document.body);

    expect(queryByTestId("file-save-dialog")).not.toBeInTheDocument();
  });

  it("should call saveFile", (): void => {
    const mockSaveFile = jest.fn();
    const { getByText } = render(
      <FileSave
        fileName="File Name"
        saveFile={mockSaveFile}
        changeFileName={jest.fn()}
      />
    );

    fireEvent.click(getByText("save.svg"));
    fireEvent.click(getByText("Save"));

    expect(mockSaveFile).toHaveBeenCalledTimes(1);
  });

  it("should call changeFileName", (): void => {
    const mockChangeFileName = jest.fn();
    const { getByText, getByPlaceholderText } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={mockChangeFileName}
      />
    );

    fireEvent.click(getByText("save.svg"));
    fireEvent.change(getByPlaceholderText("File name"), {
      target: { value: "new value" },
    });

    expect(mockChangeFileName).toHaveBeenCalledTimes(1);
  });

  it("should display error message when file name is empty", (): void => {
    const { getByText, getByPlaceholderText } = render(
      <FileSave
        fileName="File Name"
        saveFile={jest.fn()}
        changeFileName={jest.fn()}
      />
    );

    fireEvent.click(getByText("save.svg"));
    fireEvent.change(getByPlaceholderText("File name"), {
      target: { value: "" },
    });

    expect(getByText("Must not be blank")).toBeInTheDocument();
  });
});
