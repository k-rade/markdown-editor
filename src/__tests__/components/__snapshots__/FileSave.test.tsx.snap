// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSave should render 1`] = `
<body>
  <div>
    <div
      class="file-save"
    >
      <button
        class="icon-button icon-button-medium"
        type="button"
      >
        <svg>
          save.svg
        </svg>
        <div
          class="icon-button-tooltip"
        >
          Save file
        </div>
      </button>
    </div>
  </div>
</body>
`;

exports[`FileSave should render with popup 1`] = `
<body>
  <div>
    <div
      class="file-save"
    >
      <button
        class="icon-button icon-button-medium"
        type="button"
      >
        <svg>
          save.svg
        </svg>
        <div
          class="icon-button-tooltip"
        >
          Save file
        </div>
      </button>
      <div
        class="file-save-dialog"
        data-testid="file-save-dialog"
      >
        <div
          class="file-save-dialog-header"
        >
          Name File
          <button
            class="icon-button icon-button-medium"
            type="button"
          >
            <svg>
              x.svg
            </svg>
          </button>
        </div>
        <div
          class="file-save-dialog-input-wrapper"
        >
          <input
            class="file-save-dialog-input-field"
            placeholder="File name"
            type="text"
            value="File Name"
          />
          
        </div>
        <div
          class="file-save-dialog-button"
        >
          <button
            class="button button-primary"
            type="button"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
